lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_firmware.py

```diff
@@ -0,0 +1,234 @@
+    return invalid, message
+
+
+def run_update_fw_from_nw_share(idrac, module):
+    """
+    Update firmware from a network share
+    Keyword arguments:
+    idrac  -- iDRAC handle
+    module -- Ansible module
+    """
+
+    msg = {}
+    msg['changed'] = False
+    msg['msg'] = {}
+    err = False
+
+    try:
+        invalid, message = _validate_catalog_file(module.params)
+        if invalid:
+            err = True
+            msg['msg'] = message
+            return msg, err
+
+        upd_share = FileOnShare(remote=module.params['share_name'] + "/" + module.params['catalog_file_name'],
+                                mount_point=module.params['share_mnt'],
+                                isFolder=False,
+                                creds=UserCredentials(
+                                    module.params['share_user'],
+                                    module.params['share_pwd'])
+                                )
+
+        idrac.use_redfish = True
+        if '12' in idrac.ServerGeneration or '13' in idrac.ServerGeneration:
+            idrac.use_redfish = False
+
+        apply_update = True
+        msg['msg'] = idrac.update_mgr.update_from_repo(upd_share,
+                                                       apply_update,
+                                                       module.params['reboot'],
+                                                       module.params['job_wait'])
+
+        if "Status" in msg['msg']:
+            if msg['msg']['Status'] == "Success":
+                if module.params['job_wait'] is True:
+                    msg['changed'] = True
+            else:
+                err = True
+
+    except Exception as e:
+        err = True
+        msg['msg'] = "Error: %s" % str(e)
+
+    return msg, err
+
+
+# Main
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+
+            # iDRAC Credentials
+            idrac_ip=dict(required=True, type='str'),
+            idrac_user=dict(required=True, type='str'),
+            idrac_pwd=dict(required=True, type='str', no_log=True),
+            idrac_port=dict(required=False, default=443, type='int'),
+
+            # Network File Share
+            share_name=dict(required=True, type='str'),
+            share_user=dict(required=False, type='str'),
+            share_pwd=dict(required=False, type='str', no_log=True),
+            share_mnt=dict(required=True, type='str'),
+
+            # Firmware update parameters
+            catalog_file_name=dict(required=False, type='str', default='Catalog.xml'),
+            reboot=dict(required=False, default=False, type='bool'),
+            job_wait=dict(required=False, default=True, type='bool')
+        ),
+
+        supports_check_mode=False)
+
+    if not HAS_OMSDK:
+        module.fail_json(msg="Dell EMC OpenManage Python SDK required for this module")
+
+    # Connect to iDRAC
+    idrac_conn = iDRACConnection(module)
+    idrac = idrac_conn.connect()
```

Comment:
This is a bad API: initializer should return a ready-to-use connection object rather than some half-baked thing somebody should not forget to poke afterwards. Please connect right in initializer.

Also, Python has a context manager protocol especially useful for working with external resources (files, DBs, services on the network.
If any exception happens before "manual" disconnect it won't get called and the connection may stay unclosed for a while on the low-level.
You should use a context manager openning the connection in initializer and/or on enter and closing it on exit.
Besides, API like this looks just beautiful:
```python
with iDRACConnection(module) as idrac:
    msg = update_firmware_from_netshare(idrac, module)
```

Reply:
updated with this recommendation